hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java

```diff
@@ -483,17 +485,22 @@ public PartitionIdSet getOwnedPartitions() {
      */
     @Override
     public void reloadOwnedPartitions() {
-        final IPartitionService partitionService = nodeEngine.getPartitionService();
+        IPartitionService partitionService = nodeEngine.getPartitionService();
         for (; ; ) {
-            final PartitionIdSet expected = ownedPartitions.get();
-            final Collection<Integer> partitions = partitionService.getMemberPartitions(nodeEngine.getThisAddress());
-            final PartitionIdSet newSet = immutablePartitionIdSet(partitionService.getPartitionCount(), partitions);
+            PartitionIdSet expected = ownedPartitions.get();
+            Collection<Integer> partitions = partitionService.getMemberPartitions(nodeEngine.getThisAddress());
+            PartitionIdSet newSet = immutablePartitionIdSet(partitionService.getPartitionCount(), partitions);
             if (ownedPartitions.compareAndSet(expected, newSet)) {
                 return;
             }
         }
     }
 
+    @Override
+    public void nullifyOwnedPartitions() {
+        ownedPartitions.set(null);
```

Comment:
What did we gain with this change? When was it an issue?

Reply:
owned partition list is only needed for map loader. But in current codebase, every partition change triggers owned-partition-list update. This PR makes this update lazy.

Reply:
It's also used in the old query engine, hence this fix -  https://github.com/hazelcast/hazelcast/pull/11471. Is the `PartitionIdSet` large in heap consumption?

Reply:
A `PartitionIdSet` wraps a JDK `BitSet` exposing a `Set<Integer>` interface along with primitive `int` methods to avoid (un-)boxing. Its size grows with partition count (1 `long` is required per 64 partitions -- not taking into account constant overhead of one `int` + one reference), so:
- for default 271 partitions, size is 40 bytes. For comparison, a `HashSet` contains a `HashMap` whose size when empty is 44 bytes (assuming compressed OOPS - 64 bytes otherwise).
- for 2000 partitions, size is 256 bytes.
- for 20000 partitions, cost is 2,504 bytes.

At higher partition counts, a sparse `Set<Integer>` will be better in terms of memory cost than the current JDK `BitSet`. We should look into replacing the JDK `BitSet` with a better variant for high partition counts.

Reply:
🙇